--- a/daemon/remote.c
+++ b/daemon/remote.c
@@ -1272,7 +1272,7 @@
         double total_up=0, total_down=0;
         char haveStr[32];
 
-	printf( "%-4s   %-4s  %10s  %-8s  %8s  %8s  %-5s  %-11s  %s\n",
+	printf( "%-4s   %-4s  %10s  %-10s  %8s  %8s  %-5s  %-11s  %s\n",
                 "ID", "Done", "Have", "ETA", "Up", "Down", "Ratio", "Status",
                 "Name");
 
@@ -1316,7 +1316,7 @@
                 else
                     errorMark = ' ';
                 printf (
-			"%4d%c  %4s  %10s  %-8s  %8.1f  %8.1f  %5s  %-11s  %s\n",
+			"%4d%c  %4s  %10s  %-10s  %8.1f  %8.1f  %5s  %-11s  %s\n",
                   (int)id, errorMark,
                     doneStr,
                     haveStr,
